src/awkward/_v2/_connect/cuda/__init__.py

```diff
@@ -0,0 +1,142 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import os
+import glob
+import math
+import sys
+
+import numpy
+
+import awkward
+
+kernel_specializations = {
+    "awkward_ListArray32_num_64": "cuda_ListArray_num<int32_t, int64_t>",
+    "awkward_ListArrayU32_num_64": "cuda_ListArray_num<uint32_t, int64_t>",
+    "awkward_ListArray64_num_64": "cuda_ListArray_num<int64_t, int64_t>",
+    "awkward_RegularArray_num_64": "cuda_RegularArray_num<int64_t>",
+    "awkward_BitMaskedArray_to_ByteMaskedArray": "cuda_BitMaskedArray_to_ByteMaskedArray",
+    "awkward_ListArray32_validity": "cuda_ListArray_validity<int32_t>",
+    "awkward_ListArrayU32_validity": "cuda_ListArray_validity<uint32_t>",
+    "awkward_ListArray64_validity": "cuda_ListArray_validity<int64_t>",
+}
+
+try:
+    import cupy
+
+    error_message = None
+
+except ModuleNotFoundError:
+    cupy = None
+    error_message = """to use {0}, you must install cupy:
+
+    pip install cupy
+
+or
+
+    conda install -c conda-forge cupy
+"""
+
+cuda_streamptr_to_contexts = {}
+error_bits = 8
+
+kernel_errors = {}
+
+kernel = None
+
+
+def populate_kernel_errors(kernel_name, cu_file):
+    import re
+
+    pattern_errstring = "// message:"
+
+    result_errstring = [_.start() for _ in re.finditer(pattern_errstring, cu_file)]
+
+    err_strings = []
+    for index in result_errstring:
+        error = cu_file[index:]
+        error = error[error.find('"') : error.find("\n")]
+        error = error.replace('"', "")
+        err_strings.append(error)
+
+    kernel_errors[kernel_name] = err_strings
+
+
+class Invocation:
+    def __init__(self, name, error_context):
+        self.name = name
+        self.error_context = error_context
+
+
+def import_cupy(name):
+    if cupy is None:
+        raise ImportError(error_message.format(name))
+    return cupy
+
+
+def initialize_cuda_kernels(cupy):
+    if cupy is not None:
+        global kernel
+
+        if kernel is None:
+            import awkward._v2._connect.cuda._kernel_signatures
+
+            cuda_src = f"#define ERROR_BITS {error_bits}\n #define MAX_NUMPY_INT {numpy.iinfo(numpy.int64).max}"
```

Comment:
Early decision to make: should the error integer be `int64` or `uint64`? That will determine what the maximum value is.

  * It's pretty certain that we'd never have more than `np.iinfo(np.int64).max` kernel calls, so we don't _need_ it to be unsigned int.
  * But this number will never be negative, and (the reason that I usually pick signed anyway) it will never be subtracted or compared with any value that might be negative. There's no reason _not_ to make it unsigned.

Whatever you do, you want to be consistent with that.

Also, the name `MAX_NUMPY_INT` is not saying what this is about. It has nothing to do with NumPy (though that's how you come up with the value) and everything to do with being the `NO_ERROR` case. The code will be more understandable if the name reflects its role.

Reply:
I decided to move ahead with `np.uint64`.